src/renderer/decoder/polygonDecoder.js

```diff
@@ -1,78 +1,41 @@
 import * as earcut from 'earcut';
+import { addLine } from './common';
 import { getFloat32ArrayFromArray } from '../../utils/util';
-import { getLineNormal, neg } from '../../utils/geometry';
 
 // If the geometry type is 'polygon' it will triangulate the polygon list (geom)
-//      geom will be a list of polygons in which each polygon will have a flat array of vertices and a list of holes indices
-//      Example:
-/*         let geom = [
-                {
-                    flat: [
-                        0.,0., 1.,0., 1.,1., 0.,1., 0.,0, //A square
-                        0.25,0.25, 0.75,0.25, 0.75,0.75, 0.25,0.75, 0.25,0.25//A small square
-                    ]
-                    holes: [5]
-                }
-            ]
+// geom will be a list of polygons in which each polygon will have a flat array of vertices and a list of holes indices
+// Example:
+/*   let geom = [{
+       flat: [
+         0.,0., 1.,0., 1.,1., 0.,1., 0.,0, //A square
+         0.25,0.25, 0.75,0.25, 0.75,0.75, 0.25,0.75, 0.25,0.25//A small square
+       ]
+       holes: [5]
+     }]
 */
 
 export function decodePolygon (geometry) {
-    let vertices = []; // Array of triangle vertices
+    let vertices = [];
     let normals = [];
-    let breakpoints = []; // Array of indices (to vertexArray) that separate each feature
+    let breakpoints = [];
     let featureIDToVertexIndex = new Map();
-    const geometryLength = geometry.length;
 
-    for (let i = 0; i < geometryLength; i++) {
```

Comment:
I'm not the author of this micro-optimization, and, in general, we should avoid optimizations without profiling before and after the change. However, it's true that this function has been found to be a hot spot when profiling real applications. The micro-optimization of the outer loops probably don't have any real impact, but I wouldn't say the same for the inner loops, and clearly, for consistency, there is a reason to apply the same micro-optimization to the outer loop. For all of these, I would leave the code as it was.

Reply:
Do you mean `geometryLength` by `geometry.length`?

Reply:
Yes

Reply:
I have tested the performance, and there is no optimization by doing that:

https://jsperf.com/arraylengthprecaching

Reply:
Agree. But I see also in your example, in general, that there is no difference between the cached and not cached length (sometimes even `without caching length` is a little bit better).